Adding support for a generalized Utxow rule

TimSheard opened the pull request:
First steps in writing a generalized Utxow rule that can be applied to the Alonzo Era.

Generalized wisFromWitnessSet (now witsFromTxWitnesses), scriptsNeeded, validateAuxiliaryData, validateScript, verifiedWits. Made a new WellFormed constraint
HasField "adHash" (Core.TxBody era) (StrictMaybe (AuxiliaryDataHash (Crypto era)))"

TimSheard committed:
Added a start on the Utxow rules.

wrote a generic version. Replaced Shelley, Allegra, and Mary
with this generic version. Next step write an Alonzo version that
calls the generice version, and then does Alonzo stuff.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/AuxiliaryData.hs:
```diff
@@ -22,7 +24,7 @@ where
 import qualified Cardano.Ledger.Core as Core
 import qualified Cardano.Ledger.Crypto as CC (Crypto)
-import Cardano.Ledger.Era (Crypto)
+-- import Cardano.Ledger.Era (Crypto)
```
can remove

TimSheard replied:
done

TimSheard committed:
Filled in the Utxow rule

Still to do: Add a new type for errors in the Utxow rule. Make some Hasfield instances

TimSheard committed:
Completed the Utxow rule, Completed Alonzo Scripts.

The Alonzo Script datatype had a stub for Plutus scripts. Imported the right type
and completed Show, Eq, PrettyA, ToCBOR and FromCBOR instances.

The Uxtow rule is now complete. Introduced AlonzoPredFail the UtxoW PredicateFailure type
for the Alonzo era. It embeds the PredicateFaliure type from the Shelley and ShelleyMa Eras.
Wrote Show, Eq, ToCBOR and FromCBOR instances. The Utxow rule now uses this type for Failures.

Generalized the ShelleyStyleWitness to lift from one PredicateFailure type to another so it
could be reused in any ERa that uses this style of PredicateFailure Embedding.

TimSheard committed:
Added missing instances, Generalized ValidateScript class.

The missing HasField instances needed for the Uxtow rule were added.
Generalized the ValidateScript class to have the method 'scriptPrefixTag'
and made a default method for 'hashScript' Had to move a few things in
the Shelley Era to make the dependencies work out.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo.hs:
```diff
@@ -56,8 +59,10 @@ instance
   validateScript = error "TODO: implement validateScript"
-  hashScript = error "TODO: implement hashScript"
+
+-- use the default method for hashScript
```
rogue comment?

TimSheard replied:
Just commenting that there is no method definitition for hashScript since we are using the default method. I made this clearer.

JaredCorduan commented on alonzo/impl/test/lib/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs:
```diff
@@ -134,7 +134,11 @@ instance Mock c => Arbitrary (Tx (AlonzoEra c)) where
-  arbitrary = frequency [(1, pure PlutusScript), (9, NativeScript <$> arbitrary)]
+  arbitrary =
+    frequency
+      [ -- (1, pure PlutusScript) -- Until we have one of these we never generate one.
```
maybe use the word `TODO` so we don't lose track of this?

TimSheard replied:
I am hoping to get things updated to the new version of plutus, so I can resolve this.

TimSheard replied:
I added a TODO, will try and get the newer version of Plutus compilng.

TimSheard replied:
OK, with alwaysSucceeds and alwaysFails I occaisionally generate one of them, so now all possible Scripts are covered.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Address.hs:
```diff
@@ -136,6 +136,7 @@ toCred ::
   Credential kr crypto
 toCred k = KeyHashObj . hashKey $ vKey k
 
+{-  THese don't seem to be called anywhere
```
let's remove them then

TimSheard replied:
I already did this.

JaredCorduan reviewed (approved):
Thank you @TimSheard , this is a ton of great work! There are a few comments that can be removed, but besides that this looks really good!

TimSheard committed:
Moved the class ValidateScript, Roundtrip tests on PredicateFailures

The ValidateScript class was moved from Shelley.Tx(..) to Cardano.Ledger.Era(..).
The field had it type changed to scriptPrefixTag :: Core.Script era -> BS.ByteString
this way the tag written can be chosen dynamically by looking at the script.

Filled in the mising CBOR instances fo the PredicateFailure for Utxos rule.
Wrote roundtrip serilaization tests for all of the new PredicateFailure type.
BIG TODO, some Alonzo era PredicateFailure type encode Scripts, and Scripts
do not have FromCBOR instances (they have FromCBOR(Annotator Script) instance.)
We will have to address this.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs:
```diff
@@ -0,0 +1,277 @@
+-- | The Predicate failure type in the Alonzo Era. It embeds the Predicate
+--   failure type of the Shelley Era, as they share some failure modes.
+data AlonzoPredFail era
+  = Embed (UtxowPredicateFailure era)
```
The term `Embed` is already overloaded, I would suggest `WrappedShelleyEraFailure`.

TimSheard replied:
good idea.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs:
```diff
@@ -0,0 +1,277 @@
+--   failure type of the Shelley Era, as they share some failure modes.
+data AlonzoPredFail era
+  = Embed (UtxowPredicateFailure era)
+  | UnRedeemableScripts [(ScriptPurpose (Crypto era), ScriptHash (Crypto era))]
```
We need to be careful of unexpected thunks here

TimSheard replied:
OK I used rnf from the NFData class to be sure t was forced before make an UnRedeemabeScripts failure. It s used in exactly one place. Making it a StrictSequence would have required changing a half dozen different functions.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs:
```diff
@@ -0,0 +1,277 @@
+  ) =>
+  AlonzoPredFail era ->
+  Encode 'Open (AlonzoPredFail era)
+encodePredFail (Embed x) = Sum Embed 0 !> E toCBOR x
```
It might be better to flatten the encoding here, since we'd like the implementation detail (that we're wrapping) to be hidden. Shouldn't need any extra boilerplate, I think?

TimSheard replied:
This is a bit involved, since the CBOR instances for (UtxowPredicateFailure era) in the Shelley era (ie. the type of x in (Embed x)) is written by hand, so making it inline would require going back and changing that. For a future time?  Even if we flattened in CBOR, it remains unflattened (and visible) in the type AlonzoPredFail.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -56,8 +56,6 @@ module Cardano.Ledger.Alonzo.Tx
     minfee,
     isNonNativeScriptAddress,
-    -- txsize,
-    -- totExunits,
```
Delete these?

TimSheard replied:
done

nc6 commented on alonzo/impl/test/test/Test/Cardano/Ledger/Alonzo/Serialisation/Tripping.hs:
```diff
@@ -70,5 +83,12 @@ tests =
       testProperty "alonzo/AuxiliaryData" $
-        trippingAnn @(AuxiliaryData (AlonzoEra C_Crypto))
+        trippingAnn @(AuxiliaryData (AlonzoEra C_Crypto)),
+      -- testProperty "alonzo/AlonzoPredFail" $
```
Delete

TimSheard replied:
No, will get this to work, when we have better handle on the FromCBOR Annontator story and predicate failures.  I will lable it with TODO and acomment about why its is commented out.

nc6 reviewed (approved):
Bunch of spurious commented lines that should go, but otherwise looks great.

TimSheard committed:
Half way through addressing comments

TimSheard committed:
Completed addressing comments, changed project file to get latest plutus code.
